CI: TribeIT#testClusterStateNodes fails with "can't update [discovery.zen.minimum_master_nodes]"

Build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-intake/974

There are several of these failures in the logs:
```
[node_sm0/AmCrGgAAQACdSFhqAAAAAA] failed to prepareCommit settings for [Updater for: {
  1>   "key" : "discovery.zen.minimum_master_nodes",
  1>   "properties" : [
  1>     "Dynamic",
  1>     "NodeScope"
  1>   ],
  1>   "is_group_setting" : false,
  1>   "default" : "-1"
  1> }]
  1> java.lang.IllegalArgumentException: illegal value can't update [discovery.zen.minimum_master_nodes] from [-1] to [3]
  1> 	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:594) ~[main/:?]
  1> 	at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.updater(AbstractScopedSettings.java:346) ~[main/:?]
  1> 	at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:166) ~[main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterService.publishAndApplyChanges(ClusterService.java:808) ~[main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterService.runTasks(ClusterService.java:633) ~[main/:?]
  1> 	at org.elasticsearch.cluster.service.ClusterService$UpdateTask.run(ClusterService.java:1117) ~[main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:544) ~[main/:?]
  1> 	at 
```

This exhibits a real issue (but I'm not sure yet why the test fails) with the `minimum_master_nodes` dynamic setting update check.

To reproduce the issue:
- Start 2 nodes
- Dynamically update cluster settings by setting `minimum_master_nodes` to 2
- Start another node
- When that third node joins the cluster, it will throw said exception because it will check if the number of master nodes in the *previous* cluster state is at least as high as the `minimum_master_nodes` setting in the new cluster state. The number of master nodes in the previous cluster state is 0 (cluster state before the node joins the cluster) and the setting is 2.

```
Caused by: java.lang.IllegalArgumentException: cannot set discovery.zen.minimum_master_nodes to more than the current master nodes count [0]
	at org.elasticsearch.discovery.zen.ZenDiscovery.lambda$new$2(ZenDiscovery.java:176) ~[main/:?]
	at org.elasticsearch.common.settings.Setting$Updater.getValue(Setting.java:592) ~[main/:?]
	... 11 more
```

What's bad about this is that the node will not apply any of the other cluster settings in that cluster state publishing round.

/cc: @bleskes

--- a/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java
+++ b/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java
@@ -170,7 +170,17 @@ public class ZenDiscovery extends AbstractLifecycleComponent implements Discover
             this::handleMinimumMasterNodesChanged, (value) -> {
                 final ClusterState clusterState = clusterService.state();
                 int masterNodes = clusterState.nodes().getMasterNodes().size();
-                if (value > masterNodes) {
+                // the purpose of this validation is to make sure that the master doesn't step down
+                // due to a change in master nodes, which also means that there is no way to revert
+                // an accidental change. Since we validate using the current cluster state (and
+                // not the one from which the settings come from) we have to be careful and only
+                // validate if the local node is already a master. Doing so all the time causes
+                // subtle issues. For example, a node that joins a cluster has no nodes in its
+                // current cluster state. When it receives a cluster state from the master with
+                // a dynamic minimum master nodes setting int it, we must make sure we don't reject
+                // it.
+
+                if (clusterState.nodes().isLocalNodeElectedMaster() && value > masterNodes) {
                     throw new IllegalArgumentException("cannot set "
                         + ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey() + " to more than the current" +
                         " master nodes count [" + masterNodes + "]");

